1-Draft/RFCnnnn-PowerShell-Configuration.md

```diff
@@ -0,0 +1,175 @@
+---
+RFC:  RFCnnnn
+Author:  Ilya Sazonov
+Status:  Draft
+SupercededBy:  N/A
+Version: 0.1
+Area: Engine
+Comments Due: 2/25/2018
+---
+
+# `PowerShell Core` configuration
+
+## Motivation
+
+Consumers, developers, and enterpsise system administrators should be able to flexibly and conveniently configure PowerShell Core and applications based on it.
+
+The public PowerShell Core configuration API allows to use unified configuration means and to manage PowerShell-based applications in a consistent manner.
+
+## Specification
+
+`PowerShell Core` is configured using the following schemes:
+
+- On Windows - Group Policy Objects (GPO), Group Policy Preferences (GPP) and settings files.
+- On Unix - settings files.
+
+The settings files have `Json` format.
+
+**Caution!** The settings files differ from `PowerShell Core` _profile_ files that are PowerShell scripts.
+
+Configuration schemes allow to customize `PowerShell Core` in the most flexible way:
+
+- Enterprise system administrators can use GPO, GPP and computer-wide settings files to apply approved configuration settings and mandatory security settings in a centralized manner. The same settings can be applied at user, application or startup levels.
+- Developers and consumers can use user, application and startup level settings files.
+
+### Settings locations
+
+`PowerShell Core` settings are grouped into `Policy settings` and `Regular settings`. Regular settings are normal configuration settings. Regular settings can be treated as default values. Policy settings is high priority and overlap regular settings. Policy settings are used by administrators to centrally manage applications.
+
+Location | Policy settings | Regular settings
+- | - | -
+File section | "PowerShell": { "PolicySettings": {...} } | "PowerShell": { "RegularSettings": {...} }
+File section | "OtherPowerShellApplication": { "PolicySettings": {...} } | "OtherPowerShellApplication": { "RegularSettings": {...} }
+Registry key | Software\Policies\PowerShellCore | Software\PowerShellCore
+Registry key | Software\Policies\PowerShellCore\OtherPowerShellApplication | Software\PowerShellCore\OtherPowerShellApplication
+
+### Priority of applying settings
+
+Because a configuration setting can be in several schemes, the setting wins according to the priority of its scheme.
+
+#### Priorities for Policy settings in descending order
+
+Scheme | Windows | Unix
+-| - | -
+GPO -> Computer Policy | HKLM\Software\Policies\PowerShellCore | /etc/powershell.config.json
+GPO -> User Policy | HKCU\Software\Policies\PowerShellCore | See `Comment A` below
+File -> Computer-Wide | %ProgramFiles%/PowerShell/powershell.config.json | /opt/Microsoft/powershell/powershell.config.json
+File -> User-Wide | %APPDATA%/powershell.config.json | ~/powershell.config.json
```

Comment:
shouldn't this be ~/.config/powershell/powershell.config.json

Reply:
On Windows %AppData% is hidden folder. So we could the same on Unix and hide `powershell.config.json` too.

Reply:
Looks good and seems more conveniently then `~/.config/powershell/powershell.config.json`.
But I don't know that is best practice on Unix.

Reply:
If `~/.powershell.config.json` is the alternative, we should avoid it I think. Polluting $HOME with config files is IMO a bad practice. And just because you don't see them when you `ls` doesn't mean they don't impact other tools. We've already commited to the XDG config paths in the codebase, why not stay consistent with that?

Reply:
Could you please share more information? What is a standard we should follow?

Reply:
[Here](https://wiki.archlinux.org/index.php/XDG_Base_Directory)'s the Arch Wiki summary article on the XDG Base Directory specification (it's more readable than the specification page and it links to it anyway).

We already use it for the [personal module path](https://github.com/PowerShell/PowerShell/blob/79f21b41de0de9b2f68a19ba1fdef0b98f3fb1cb/src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs#L940-L951), [user drive](https://github.com/PowerShell/PowerShell/blob/c5dd3bd2c94ad59fdd66a6752cb477daf2cb7d40/src/System.Management.Automation/engine/InitialSessionState.cs#L2664-L2676) (not sure I know what that is), the [analysis cache](https://github.com/PowerShell/PowerShell/blob/8eebc2699c067882d5558f84d068a805d94e432f/src/System.Management.Automation/engine/Modules/AnalysisCache.cs#L1115-L1119), and the [current user configuration location](https://github.com/PowerShell/PowerShell/blob/c5dd3bd2c94ad59fdd66a6752cb477daf2cb7d40/src/System.Management.Automation/engine/Utils.cs#L234-L246).

Reply:
Thanks! So we use `XDG_CONFIG_HOME` with default `$HOME/.config`.